File: scripts/make_icon.py
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont
import io
import shutil
import struct
import subprocess

root = Path(__file__).resolve().parents[1]
iconset = root / "Resources" / "AppIcon.iconset"
iconset.mkdir(parents=True, exist_ok=True)
icns_path = root / "Resources" / "AppIcon.icns"

for old_file in iconset.glob("*.png"):
    old_file.unlink()

sizes = [16, 32, 128, 256, 512]

def draw_icon(size: int) -> Image.Image:
    scale = size / 1024
    image = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    draw = ImageDraw.Draw(image)
    radius = int(210 * scale)

    for y in range(size):
        t = y / max(size - 1, 1)
        x_mix = 0.18
        color = (
            int(52 + 100 * t + 24 * x_mix),
            int(173 - 76 * t),
            int(240 - 12 * t),
            255,
        )
        draw.line([(0, y), (size, y)], fill=color)

    outer_mask = Image.new("L", (size, size), 0)
    ImageDraw.Draw(outer_mask).rounded_rectangle([0, 0, size - 1, size - 1], radius=radius, fill=255)
    image.putalpha(outer_mask)

    glow = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    glow_draw = ImageDraw.Draw(glow)
    glow_draw.ellipse([int(-150 * scale), int(-110 * scale), int(590 * scale), int(560 * scale)], fill=(108, 223, 255, 78))
    glow_draw.ellipse([int(360 * scale), int(360 * scale), int(1190 * scale), int(1160 * scale)], fill=(131, 82, 235, 95))
    image.alpha_composite(glow)

    inset = int(150 * scale)
    panel = [inset, inset, size - inset, size - inset]
    draw = ImageDraw.Draw(image)
    draw.rounded_rectangle(panel, radius=int(150 * scale), fill=(21, 28, 48, 245))
    draw.rounded_rectangle(
        [panel[0] + int(24 * scale), panel[1] + int(24 * scale), panel[2] - int(24 * scale), panel[3] - int(24 * scale)],
        radius=int(126 * scale),
        outline=(255, 255, 255, 25),
        width=max(1, int(5 * scale)),
    )

    try:
        font = ImageFont.truetype("/System/Library/Fonts/Supplemental/Arial Bold.ttf", int(485 * scale))
    except Exception:
        font = ImageFont.load_default()
    text = "Q"
    bbox = draw.textbbox((0, 0), text, font=font)
    tx = (size - (bbox[2] - bbox[0])) / 2 - int(7 * scale)
    ty = (size - (bbox[3] - bbox[1])) / 2 - int(70 * scale)
    draw.text((tx, ty), text, fill=(250, 253, 255, 255), font=font)

    badge = int(255 * scale)
    draw.ellipse(
        [size - badge - int(24 * scale), size - badge - int(24 * scale), size + int(78 * scale), size + int(78 * scale)],
        fill=(255, 202, 68, 255),
    )
    draw.ellipse(
        [size - int(170 * scale), size - int(170 * scale), size - int(92 * scale), size - int(92 * scale)],
        fill=(255, 232, 142, 210),
    )
    return image

base = draw_icon(1024)
for size in sizes:
    img = base.resize((size, size), Image.Resampling.LANCZOS)
    img.convert("RGB").save(iconset / f"icon_{size}x{size}.png")
    retina = base.resize((size * 2, size * 2), Image.Resampling.LANCZOS)
    retina.convert("RGB").save(iconset / f"icon_{size}x{size}@2x.png")

if shutil.which("iconutil"):
    result = subprocess.run(
        ["iconutil", "-c", "icns", str(iconset), "-o", str(icns_path)],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        check=False,
    )
    if result.returncode == 0:
        raise SystemExit(0)

chunks = [
    ("icp4", 16),
    ("icp5", 32),
    ("icp6", 64),
    ("ic07", 128),
    ("ic08", 256),
    ("ic09", 512),
    ("ic10", 1024),
]

payload = bytearray()
for chunk_type, size in chunks:
    img = base.resize((size, size), Image.Resampling.LANCZOS)
    png = io.BytesIO()
    img.save(png, format="PNG")
    data = png.getvalue()
    payload.extend(chunk_type.encode("ascii"))
    payload.extend(struct.pack(">I", len(data) + 8))
    payload.extend(data)

icns = bytearray()
icns.extend(b"icns")
icns.extend(struct.pack(">I", len(payload) + 8))
icns.extend(payload)
icns_path.write_bytes(icns)

File: scripts/make_icon.ts
import { createCanvas, Canvas } from "@napi-rs/canvas";
import { execFile } from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { promisify } from "node:util";

const execFileAsync = promisify(execFile);

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const root = path.resolve(__dirname, "..");
const resourcesDir = path.join(root, "Resources");
const iconsetDir = path.join(resourcesDir, "AppIcon.iconset");
const icnsPath = path.join(resourcesDir, "AppIcon.icns");

const sizes = [16, 32, 128, 256, 512];

async function pathExists(targetPath: string): Promise<boolean> {
  try {
    await fs.access(targetPath);
    return true;
  } catch {
    return false;
  }
}

export function drawIcon(size: number): Canvas {
  const scale = size / 1024;
  const canvas = createCanvas(size, size);
  const ctx = canvas.getContext("2d");

  const radius = Math.round(210 * scale);

  // Outer rounded rectangle clip
  ctx.save();
  ctx.beginPath();
  ctx.roundRect(0, 0, size, size, radius);
  ctx.clip();

  // Vertical gradient background
  const bgGrad = ctx.createLinearGradient(0, 0, 0, size);
  bgGrad.addColorStop(0, "rgb(56, 173, 240)");
  bgGrad.addColorStop(1, "rgb(156, 97, 228)");
  ctx.fillStyle = bgGrad;
  ctx.fillRect(0, 0, size, size);

  // Glow ellipse 1
  ctx.beginPath();
  ctx.ellipse(
    220 * scale,
    225 * scale,
    370 * scale,
    335 * scale,
    0,
    0,
    Math.PI * 2
  );
  ctx.fillStyle = "rgba(108, 223, 255, 0.306)";
  ctx.fill();

  // Glow ellipse 2
  ctx.beginPath();
  ctx.ellipse(
    775 * scale,
    760 * scale,
    415 * scale,
    400 * scale,
    0,
    0,
    Math.PI * 2
  );
  ctx.fillStyle = "rgba(131, 82, 235, 0.373)";
  ctx.fill();

  // Center panel
  const inset = 150 * scale;
  const panelSize = size - 2 * inset;
  ctx.beginPath();
  ctx.roundRect(inset, inset, panelSize, panelSize, 150 * scale);
  ctx.fillStyle = "rgba(21, 28, 48, 0.961)";
  ctx.fill();

  // Panel inner border
  const borderOffset = inset + 24 * scale;
  const borderSize = size - 2 * borderOffset;
  ctx.beginPath();
  ctx.roundRect(borderOffset, borderOffset, borderSize, borderSize, 126 * scale);
  ctx.lineWidth = Math.max(1, 5 * scale);
  ctx.strokeStyle = "rgba(255, 255, 255, 0.098)";
  ctx.stroke();

  // Letter "Q"
  const fontSize = Math.round(485 * scale);
  ctx.font = `bold ${fontSize}px "Arial Bold", Arial, -apple-system, sans-serif`;
  ctx.fillStyle = "rgba(250, 253, 255, 1.0)";
  ctx.textAlign = "center";
  ctx.textBaseline = "middle";
  ctx.fillText("Q", size / 2 - 7 * scale, size / 2 - 28 * scale);

  // Badge outer circle
  ctx.beginPath();
  ctx.arc(923.5 * scale, 923.5 * scale, 178.5 * scale, 0, Math.PI * 2);
  ctx.fillStyle = "rgba(255, 202, 68, 1.0)";
  ctx.fill();

  // Badge inner circle
  ctx.beginPath();
  ctx.arc(893 * scale, 893 * scale, 39 * scale, 0, Math.PI * 2);
  ctx.fillStyle = "rgba(255, 232, 142, 0.824)";
  ctx.fill();

  ctx.restore();

  return canvas;
}

export async function buildIcnsBinary(
  baseCanvas: Canvas,
  targetPath: string
): Promise<void> {
  const chunks: Array<{ type: string; size: number }> = [
    { type: "icp4", size: 16 },
    { type: "icp5", size: 32 },
    { type: "icp6", size: 64 },
    { type: "ic07", size: 128 },
    { type: "ic08", size: 256 },
    { type: "ic09", size: 512 },
    { type: "ic10", size: 1024 },
  ];

  const payloadChunks = await Promise.all(
    chunks.map(async (chunk) => {
      const resizedCanvas = createCanvas(chunk.size, chunk.size);
      const ctx = resizedCanvas.getContext("2d");
      ctx.drawImage(baseCanvas, 0, 0, chunk.size, chunk.size);
      const pngBuffer = await resizedCanvas.encode("png");

      const header = Buffer.alloc(8);
      header.write(chunk.type, 0, 4, "ascii");
      header.writeUInt32BE(pngBuffer.length + 8, 4);

      return Buffer.concat([header, pngBuffer]);
    })
  );

  const payload = Buffer.concat(payloadChunks);
  const fileHeader = Buffer.alloc(8);
  fileHeader.write("icns", 0, 4, "ascii");
  fileHeader.writeUInt32BE(payload.length + 8, 4);

  const finalBuffer = Buffer.concat([fileHeader, payload]);
  await fs.writeFile(targetPath, finalBuffer);
}

export async function generateIcons(): Promise<void> {
  console.log("Generating app icons...");
  await fs.mkdir(iconsetDir, { recursive: true });

  // Clear existing pngs in iconset asynchronously
  const files = await fs.readdir(iconsetDir);
  await Promise.all(
    files
      .filter((file) => file.endsWith(".png"))
      .map((file) => fs.unlink(path.join(iconsetDir, file)))
  );

  const baseCanvas = drawIcon(1024);

  // Generate multi-scale PNG icons concurrently
  await Promise.all(
    sizes.flatMap((size) => {
      // 1x
      const task1x = (async () => {
        const canvas1x = createCanvas(size, size);
        const ctx1x = canvas1x.getContext("2d");
        ctx1x.drawImage(baseCanvas, 0, 0, size, size);
        const buffer1x = await canvas1x.encode("png");
        await fs.writeFile(
          path.join(iconsetDir, `icon_${size}x${size}.png`),
          buffer1x
        );
      })();

      // 2x Retina
      const task2x = (async () => {
        const retinaSize = size * 2;
        const canvas2x = createCanvas(retinaSize, retinaSize);
        const ctx2x = canvas2x.getContext("2d");
        ctx2x.drawImage(baseCanvas, 0, 0, retinaSize, retinaSize);
        const buffer2x = await canvas2x.encode("png");
        await fs.writeFile(
          path.join(iconsetDir, `icon_${size}x${size}@2x.png`),
          buffer2x
        );
      })();

      return [task1x, task2x];
    })
  );

  // Generate ICNS using iconutil if available, otherwise pure async binary
  let generatedViaIconutil = false;
  try {
    await execFileAsync("iconutil", ["-c", "icns", iconsetDir, "-o", icnsPath]);
    generatedViaIconutil = true;
  } catch {
    generatedViaIconutil = false;
  }

  if (!generatedViaIconutil) {
    await buildIcnsBinary(baseCanvas, icnsPath);
  }

  console.log(`Icons successfully generated at ${icnsPath}`);
}

if (process.argv[1] && path.resolve(process.argv[1]) === __filename) {
  await generateIcons();
}
